src/Datadog.Trace.DuckTyping/ILHelpers.cs

```diff
@@ -0,0 +1,307 @@
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace Datadog.Trace.DuckTyping
+{
+    /// <summary>
+    /// Internal IL Helpers
+    /// </summary>
+    internal static class ILHelpers
+    {
+        private static Func<DynamicMethod, RuntimeMethodHandle> _dynamicGetMethodDescriptor;
+
+        /// <summary>
+        /// Load instance argument
+        /// </summary>
+        /// <param name="il">ILGenerator</param>
+        /// <param name="actualType">Actual type</param>
+        /// <param name="expectedType">Expected type</param>
+        internal static void LoadInstanceArgument(ILGenerator il, Type actualType, Type expectedType)
+        {
+            il.Emit(OpCodes.Ldarg_0);
+            if (actualType == expectedType)
+            {
+                return;
+            }
+
+            if (expectedType.IsValueType)
+            {
+                il.DeclareLocal(expectedType);
+                il.Emit(OpCodes.Unbox_Any, expectedType);
```

Comment:
What I'm trying to do with that code is similar to `LoadObjectAsArgument` here: 
https://sharplab.io/#v2:EYLgHgbALANALiAlgGwD4AEBMBGAsAKHQGYACLEgYRIG8CT6zT1sIyoSAZAewEMATAPLAAVgFMAxnACCAZykAnAOYBXALaiAdnAAUXERLglEGmXB4bxogJQ06De9u0BZAJ4A1HsmWiAKi4AO1sam5pZWAHQAyjwuABKI2lYA3Hb0AL6pJJmm8sqSJK4eXr4Borb49vbEbCTRcQk2tBWVLQD0rSSxosjIXCQA6lzyyHwAhJn2Gc3pBGlAA===

That applies when the value type is non public, then the proxy stores it as an object to bypass visibility checks, and it's pass it to the dynamic method as it. Because the dynamic method can disable visibility checks I unbox it from there, and call the inner members.